File: C#/ToyLanguage/ToyLanguage/Model/Collections/WrapperDictionary.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ToyLanguage.Interfaces;

namespace ToyLanguage.Model.Collections
{
    [Serializable]
    class WrapperDictionary<K, V> : IMyDictionary<K, V>
    {
        private Dictionary<K, V> map;
            
            public WrapperDictionary()
        {
            map = new Dictionary<K, V>();
        } 

        public void addAll(Dictionary<K, V> all)
        {
            map.Clear();
            map = new Dictionary<K, V>(all);
        }

        public Dictionary<K, V> getMap()
        {
            return map;
        }

        public V lookUp(K id)
        {
            return map[id];
        }

        public bool isEmpty()
        {
            if(map.Count == 0)
            {
                return true;
            }
            return false;
        }

        public int size()
        {
            return map.Count;
        }

        public bool containsKey(K key)
        {
            return map.ContainsKey(key);
        }

        public void put(K key, V value)
        {
            if (map.ContainsKey(key))
            {
                map[key] = value;
            }
            else
            {
                map.Add(key, value);
            }
            
        }

        public void remove(K key)
        {
            map.Remove(key);
        }

        public string ToString()
        {
            StringBuilder result = new StringBuilder();

            foreach (KeyValuePair<K, V> entry in map)
            {
                result.Append(entry.Key + "->" + entry.Value + Environment.NewLine);
            }
            return result.ToString();
        }

    }
}

File: Java/src/model/Collections/WrapperDictionary.java
package model.Collections;

import interfaces.IDictionary;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by Lucian on 11/2/2015.
 */
public class WrapperDictionary<K, V> implements IDictionary<K, V> {

    /**
     * The storage map
     */
    protected Map<K, V> map;

    /**
     * The constructor
     */
    public WrapperDictionary() {
        this.map = new HashMap<>();
    }

    /**
     * Get the value of the id
     *
     * @param id The id
     * @return Value
     */
    @Override
    public V lookUp(K id) {
        return map.get(id);
    }

    /**
     * Check if dictionary is empty
     *
     * @return True if yes, false otherwise
     */
    @Override
    public boolean isEmpty() {
        return map.isEmpty();
    }

    /**
     * Get the size of the dictionary
     *
     * @return int size
     */
    @Override
    public int size() {
        return map.size();
    }

    /**
     * Check if K key is in dictionary
     *
     * @param key The key
     * @return True if yes, false otherwise
     */
    @Override
    public boolean containsKey(K key) {
        return map.containsKey(key);
    }

    /**
     * Put the V value in K key
     *
     * @param key   The key
     * @param value The value
     */
    @Override
    public void put(K key, V value) {
        map.put(key, value);
    }

    /**
     * Remove the key from dictionary
     *
     * @param key The key
     */
    @Override
    public void remove(K key) {
        map.remove(key);
    }

    /**
     * String representation
     *
     * @return String
     *
     */
    @Override
    public String toString() {
        StringBuilder result = new StringBuilder();
        for (Map.Entry<K, V> entry : map.entrySet())
        {
            result.append(entry.getKey() + "->" + entry.getValue() + "\n");
        }

        return result.toString();
    }

    @Override
    public void addAll(Map<K, V> all) {
        map.clear();
        map.putAll(all);
    }

    @Override
    public Map<K, V> getMap() {
        return map;
    }
}
